File: src/inputs/plugins/air_quality/connector/aqicn.py
import asyncio

import aiohttp

from inputs.plugins.air_quality.connector.base import (
    AirQualityConnector,
    AirQualityData,
)


class AqicnConnector(AirQualityConnector):
    """
    Air quality connector using AQICN (World Air Quality Index) API.

    Fetches real-time AQI and pollutant data from waqi.info.
    Use this connector when no physical sensor is available.

    API docs: https://aqicn.org/json-api/doc/
    Free token: https://aqicn.org/data-platform/token/
    """

    def __init__(self, config: dict):
        """
        Parameters
        ----------
        config : dict
            Must contain:
            - api_key (str): AQICN token
            - latitude (float): location latitude
            - longitude (float): location longitude
        """
        super().__init__(config)
        self.api_key: str = config.get("api_key", "")
        self.latitude: float = config.get("latitude", -6.2088)
        self.longitude: float = config.get("longitude", 106.8456)

    async def connect(self) -> bool:
        """Validate API key and confirm connector is ready."""
        if not self.api_key:
            self.logger.warning("AqicnConnector: no API key provided")
            return False
        return True

    async def disconnect(self) -> None:
        """No-op: stateless HTTP connector requires no teardown."""
        pass  # Stateless HTTP — nothing to close

    async def read(self) -> AirQualityData | None:
        """
        Fetch air quality data from AQICN API.

        Returns
        -------
        AirQualityData or None
            Parsed data, or None if request failed.
        """
        if not self.api_key:
            return None

        url = f"https://api.waqi.info/feed/geo:{self.latitude};{self.longitude}/"
        params = {"token": self.api_key}

        try:
            timeout = aiohttp.ClientTimeout(total=10)
            async with aiohttp.ClientSession(timeout=timeout) as session:
                async with session.get(url, params=params) as response:
                    if response.status != 200:
                        self.logger.error(
                            f"AqicnConnector: HTTP {response.status}: {await response.text()}"
                        )
                        return None

                    payload = await response.json()

            if payload.get("status") != "ok":
                self.logger.error(f"AqicnConnector: API error: {payload.get('data')}")
                return None

            return self._parse(payload)

        except asyncio.TimeoutError:
            self.logger.error("AqicnConnector: request timed out")
            return None
        except aiohttp.ClientError as e:
            self.logger.error(f"AqicnConnector: network error: {e}")
            return None
        except Exception as e:
            self.logger.error(f"AqicnConnector: unexpected error: {e}")
            return None

    def _parse(self, payload: dict) -> AirQualityData:
        """
        Parse AQICN API response into AirQualityData.

        Parameters
        ----------
        payload : dict
            Raw API response with status == 'ok'.

        Returns
        -------
        AirQualityData
        """
        data = payload.get("data", {})
        iaqi = data.get("iaqi", {})

        def get_val(key: str) -> float | None:
            entry = iaqi.get(key, {})
            return entry.get("v") if entry else None

        aqi_raw = data.get("aqi", "-")
        aqi = int(aqi_raw) if aqi_raw not in ("-", None) else None

        location = data.get("city", {}).get("name", "Unknown")

        return AirQualityData(
            aqi=aqi,
            pm25=get_val("pm25"),
            pm10=get_val("pm10"),
            co=get_val("co"),
            no2=get_val("no2"),
            so2=get_val("so2"),
            o3=get_val("o3"),
            temperature=get_val("t"),
            humidity=get_val("h"),
            location=location,
            source="aqicn",
        )

File: plugins/inputs/air_quality/connector/aqicn/aqicn.go
package aqicn

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/openmind/om1/plugins/inputs/air_quality/connector"
)

// Config mirrors the `config: dict` passed into Python's AqicnConnector.__init__.
type Config struct {
	APIKey    string  // config["api_key"]
	Latitude  float64 // config["latitude"], default -6.2088
	Longitude float64 // config["longitude"], default 106.8456
}

// Connector implements connector.AirQualityConnector for the AQICN cloud API.
// Mirrors Python's AqicnConnector.
type Connector struct {
	cfg        Config
	httpClient *http.Client
	logger     *log.Logger
}

// New creates a new AQICN connector, applying the same defaults as the Python version.
func New(cfg Config, logger *log.Logger) *Connector {
	if cfg.Latitude == 0 && cfg.Longitude == 0 {
		cfg.Latitude = -6.2088
		cfg.Longitude = 106.8456
	}
	if logger == nil {
		logger = log.Default()
	}
	return &Connector{
		cfg:        cfg,
		httpClient: &http.Client{Timeout: 10 * time.Second},
		logger:     logger,
	}
}

func (c *Connector) Name() string {
	return "aqicn"
}

// Connect validates the API key and confirms the connector is ready.
// Mirrors Python's `connect()`.
func (c *Connector) Connect(ctx context.Context) (bool, error) {
	if c.cfg.APIKey == "" {
		c.logger.Println("AqicnConnector: no API key provided")
		return false, nil
	}
	return true, nil
}

// Disconnect is a no-op: stateless HTTP connector requires no teardown.
func (c *Connector) Disconnect(ctx context.Context) error {
	return nil
}

type iaqiEntry struct {
	V *float64 `json:"v"`
}

type aqicnPayload struct {
	Status string `json:"status"`
	Data   struct {
		AQI  json.Number `json:"aqi"`
		City struct {
			Name string `json:"name"`
		} `json:"city"`
		IAQI struct {
			PM25 *iaqiEntry `json:"pm25"`
			PM10 *iaqiEntry `json:"pm10"`
			CO   *iaqiEntry `json:"co"`
			NO2  *iaqiEntry `json:"no2"`
			SO2  *iaqiEntry `json:"so2"`
			O3   *iaqiEntry `json:"o3"`
			T    *iaqiEntry `json:"t"`
			H    *iaqiEntry `json:"h"`
		} `json:"iaqi"`
	} `json:"data"`
}

// Read fetches air quality data from the AQICN API.
// Returns (nil, nil) on any handled failure, mirroring Python's `return None`.
func (c *Connector) Read(ctx context.Context) (*connector.AirQualityData, error) {
	if c.cfg.APIKey == "" {
		return nil, nil
	}

	url := fmt.Sprintf("https://api.waqi.info/feed/geo:%v;%v/?token=%s", c.cfg.Latitude, c.cfg.Longitude, c.cfg.APIKey)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		c.logger.Printf("AqicnConnector: unexpected error: %v", err)
		return nil, nil
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		c.logger.Printf("AqicnConnector: network error: %v", err)
		return nil, nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		c.logger.Printf("AqicnConnector: HTTP %d", resp.StatusCode)
		return nil, nil
	}

	var payload aqicnPayload
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		c.logger.Printf("AqicnConnector: unexpected error: %v", err)
		return nil, nil
	}

	if payload.Status != "ok" {
		c.logger.Printf("AqicnConnector: API error, status=%q", payload.Status)
		return nil, nil
	}

	return c.parse(payload), nil
}

// parse mirrors Python's `_parse(payload)`.
func (c *Connector) parse(payload aqicnPayload) *connector.AirQualityData {
	data := connector.NewAirQualityData()

	// aqi_raw = data.get("aqi", "-"); aqi = int(aqi_raw) if aqi_raw not in ("-", None) else None
	if aqiFloat, err := payload.Data.AQI.Float64(); err == nil {
		aqi := int(aqiFloat)
		data.AQI = &aqi
	}

	get := func(e *iaqiEntry) *float64 {
		if e == nil {
			return nil
		}
		return e.V
	}

	data.PM25 = get(payload.Data.IAQI.PM25)
	data.PM10 = get(payload.Data.IAQI.PM10)
	data.CO = get(payload.Data.IAQI.CO)
	data.NO2 = get(payload.Data.IAQI.NO2)
	data.SO2 = get(payload.Data.IAQI.SO2)
	data.O3 = get(payload.Data.IAQI.O3)
	data.Temperature = get(payload.Data.IAQI.T)
	data.Humidity = get(payload.Data.IAQI.H)

	if payload.Data.City.Name != "" {
		data.Location = payload.Data.City.Name
	} else {
		data.Location = "Unknown"
	}
	data.Source = "aqicn"

	return data
}
